[WIP] Opaque result types prototype

jckarter opened the pull request:
WIP towards a prototype of an MVP subset of https://forums.swift.org/t/opaque-result-types/15645 . For an initial implementation, this won't include the resilience, second-order `where` constraint, or multiple conditional conformance features proposed in the thread.

- [x] Parsing
- [x] Decl representation for opaque types
- [x] Type representation for opaque types (probably as some new kind of ArchetypeType?)
- [x] Type checking to resolve the underlying type of opaque types
- [x] Non-resilient SILGen (lowering the opaque type directly to its underlying type)
- [x] Serialization support

Limitations include:

- The opaque type is spelled `__opaque Protocol` or `__opaque ProtocolA & ProtocolB [& ProtocolC...]` as placeholder spelling.
- Opaque types can only appear as the return type of `func` declarations, not properties or closures.
- `where` clause constraints on the opaque type are not yet supported, nor are conditional conformances on the opaque type. However, the protocols in the `__opaque` type can be arbitrary protocols, with any number of associated type or Self type requirements. `__opaque` types can in turn be inferred as associated types of protocol conformances.
- The runtime and resilience aspects are not yet implemented. Changing the underlying type of a public opaque type will break ABI.

Some known issues:

- Because of the missing runtime support, I substitute out opaque types during SILGen in a rather hacky unprincipled way. It's likely that the optimizer, particularly generic specialization, may still cause crashes.
- The type checker currently crashes if there's a type error in the underlying type, such as if it does not conform to the required protocols.
- Diagnostics involving opaque types will print the type with a compiler-internal representation involving the original decl name and generic arguments, instead of anything human-understandable.

jckarter committed:
Parse opaque types.

jckarter committed:
AST: Represent OpaqueTypeDecls.

To represent the abstracted interface of an opaque type, we need a generic signature that refines
the outer context generic signature with an additional generic parameter representing the underlying
type and its exposed constraints. Opaque types also need to be keyed by their originating decl, so
that we can treat values of the same opaque type as the same. When we check a FuncDecl with an
opaque type specified as its return type, create an OpaqueTypeDecl and associate it with the
originating decl. (A representation for *types* derived from the opaque decl will come next.)

jckarter committed:
Add an OpaqueTypeArchetypeType subclass.

DougGregor commented on lib/AST/ASTContext.cpp:
```diff
@@ -4270,6 +4272,64 @@ DependentMemberType *DependentMemberType::get(Type base,
+  auto newOpaque = std::move(builder)
+    .computeOpaqueArchetype(Decl, Substitutions, properties, arena);
+  
+  set->InsertNode(newOpaque, insertPos);
```
Please perform another `FindNodeOrInsertPos` here. Some other code might have perturbed the set, invalidating `insertPos`.

DougGregor commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -5210,6 +5221,26 @@ ConstraintResult GenericSignatureBuilder::addSameTypeRequirement(
                                       diagnoseMismatch);
 }
 
+ConstraintResult GenericSignatureBuilder::addSameTypeRequirementToConcrete(
```
I don't think you need this if you take my `addRequirement` suggestion.

DougGregor commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -7518,6 +7549,31 @@ GenericSignature *GenericSignatureBuilder::computeRequirementSignature(
+  Type superclass;
+  LayoutConstraint layout;
+  SmallVector<ProtocolDecl*, 4> protos;
+  equivClass->getArchetypeConstraints(protos, superclass, layout);
```
Rather than calling `finalize` / `resolveEquivalenceClass` / `getArchetypeConstraints`, how about calling `computeGenericSignature()` and then `GenericSignature::getConformsTo()` / `::getSuperclassBound()` / `::getLayoutConstraint()` ? The `GenericSignature` is going to get created at some point anyway.

jckarter replied:
I'm trying to avoid forming a GenericSignature here, because it will be invalid if there are bindings to generic parameters outside of the signature.

DougGregor commented on include/swift/AST/Types.h:
```diff
@@ -4757,6 +4759,64 @@ class PrimaryArchetypeType final : public ArchetypeType,
+  OpaqueTypeArchetypeType(OpaqueTypeDecl *OpaqueDecl,
+                          SubstitutionMap Substitutions,
+                          RecursiveTypeProperties Props,
+                          GenericSignatureBuilder &&Builder,
```
I'd prefer that we pass the `GenericSignature` itself (whose queries will eventually go through the GSB)` rather than the `GenericSignatureBuilder`. We want the GSB instances to be mostly invisible to the rest of the AST.

DougGregor commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -2293,6 +2293,21 @@ Type EquivalenceClass::getAnchor(
 }
 
+void
+EquivalenceClass::getArchetypeConstraints(SmallVectorImpl<ProtocolDecl *> &protos,
```
My suggestions below will make this unnecessary.

DougGregor commented on lib/AST/Type.cpp:
```diff
@@ -1190,6 +1191,13 @@ CanType TypeBase::computeCanonicalType() {
+    // Canonicalize the substitution map.
+    OpaqueTypeArchetypeType *OAT = cast<OpaqueTypeArchetypeType>(this);
+    auto canonicalSubs = OAT->getSubstitutions().getCanonical();
+    Result = OpaqueTypeArchetypeType::get(OAT->getOpaqueDecl(), canonicalSubs);
```
This is going to end up creating a new GSB instance and finalizing it and such. Can we canonicalize all of the individual pieces and call into the other, more direct constructor that won't do that redundant work?

DougGregor replied:
Hmm. There's another issue here. `ArchetypeType` is a bit special because they've always been considered canonical by construction. That's why the `ArchetypeType` constructor takes its `ASTContext` by reference rather than by pointer. Types that have non-canonical forms take `ASTContext` pointers where NULL indicates that the type is not already canonical. I think you'll need to change the `ArchetypeType` constructor to take the `ASTContext` as a pointer, and `OpaqueTypeArchetypeType::get` will determine whether it's forming the canonical type to decide whether to pass down the context or NULL.

DougGregor commented on lib/AST/Type.cpp:
```diff
@@ -3532,6 +3566,35 @@ case TypeKind::Id:
+    auto newSubMap =
+      SubstitutionMap::get(opaque->getOpaqueDecl()->getGenericSignature(),
+                           newSubs,
+                           opaque->getSubstitutions().getConformances());
```
If the transform is doing anything interesting, passing along conformances will be wrong. It's probably safest to use `SubstitutionMap::get(the-generic-signature, a-function-that-uses-the-replaced-types, LookupConformanceInModule(module-of-the-opaque-decl)`.

jckarter replied:
Thanks, I was aware of that problem but wasn't sure how to force it to re-lookup the conformances. It seems like `transformRec` really ought to be working with something akin to the old `Substitution` record instead of types, though.

jckarter commented on lib/AST/ASTContext.cpp:
```diff
@@ -4270,6 +4272,64 @@ DependentMemberType *DependentMemberType::get(Type base,
+  // replacements in the substitution map.
+  for (auto outerParam : Decl->getGenericSignature()->getGenericParams()) {
+    auto boundType = Type(outerParam).subst(Substitutions);
+    builder.addSameTypeRequirementToConcrete(Type(outerParam),
```
If `boundType` is itself a `GenericTypeParamType` or `DependentMemberType`, then that will try to map it to the equivalence class for the argument inside the GenericSignatureBuilder, won't it? I specifically don't want the second type to be treated as "inside" the generic signature here, but to always be treated as a concrete type.

jckarter committed:
Sema: Infer the underlying type for opaque return types from function bodies.

xedin commented on lib/Sema/CSSolver.cpp:
```diff
@@ -1115,6 +1115,11 @@ ConstraintSystem::solveImpl(Expr *&expr,
     auto constraintKind = ConstraintKind::Conversion;
+    
+    if (getContextualTypePurpose() == CTP_ReturnStmt
+        && Options.contains(ConstraintSystemFlags::UnderlyingTypeForOpaqueReturnType))
```
I wonder - do we really need a new flag or we could detect this situation based on `convertType` itself, because it always would be `OpaqueTypeArchetypeType` right?

jckarter replied:
I think we normally want to treat the opaque type and underlying type as distinct outside of return statements.

xedin replied:
Sorry, what I meant was:

```c++
if (getContextualTypePurpose() == CTP_ReturnStmt &&
    covertType->isa<OpaqueTypeArchetypeType>())
  constraintKind = ConstraintKind::OpaqueUnderlyingType;
```

jckarter replied:
The docs for ContextualTypePurpose suggested that it was only intended as a discriminator for diagnostics, so I wasn't sure whether it was OK to key other constraint system behavior solely on it. I can do what you suggested if you think it's better.

xedin replied:
I think it was intended like that originally but it would be totally okay to use it for this case as well to avoid adding new flags.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3925,6 +3930,42 @@ ConstraintSystem::simplifyDynamicTypeOfConstraint(
+    }
+  }
+
+  addConstraint(ConstraintKind::Equal, type1, underlyingTyVar, locator);
```
Just for my better understanding, could you please elaborate if this constraint has to be equality, or it could be conversion instead? It seems like in the non-opaque case solver creates a conversion constraint from contextual result to what we infer for `return`, but in this case I _think_ equality makes more sense because base is going to be a type variable with a set of requirements, is that correct? If so, I wonder if we should just bind these types instead of to pull requirements to the left type?

xedin replied:
bq. Are there cases where a Conversion constraint between type variables might still allow a more general set of types to be bound?

I can only think of implicit conversions being a difference between equals and conversion in that regard.

bq. Bind seems too specific since we still want to allow lvalues in the return statement to be loaded too.

Actually if you look at the `matchTypes` - both equal and bind share the same path and most likely going to bind types or merge equivalence classes of the type variables if either is a type variable.

xedin reviewed (commented):
Solver changes look reasonable to me, @jckarter!

jckarter committed:
SILGen: Substitute away opaque types.

jckarter committed:
Support nested types on opaque archetypes (and maybe opened ones).

jckarter committed:
Serialization: Serialize opaque types and their decls.

jckarter committed:
IRGen: Substitute out opaque types when emitting associated types.

jckarter committed:
Enable opaque result types.

jckarter commented:
@swift-ci Please Build Toolchain macOS Platform

jckarter commented:
@swift-ci Please build toolchain macOS platform

jckarter commented:
@swift-ci Please build toolchain macOS platform

slavapestov commented on include/swift/SIL/AbstractionPattern.h:
```diff
@@ -298,9 +298,15 @@ class AbstractionPattern {
     assert(signature || !origType->hasTypeParameter());
     TheKind = unsigned(kind);
-    OrigType = origType;
+    // If origType is already
```
already what

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -5703,8 +5740,10 @@ void AbstractFunctionDecl::computeType(AnyFunctionType::ExtInfo info) {
     resultTy = fn->getBodyResultTypeLoc().getType();
-    if (!resultTy)
+    if (!resultTy) {
+      
```
Blank line

slavapestov commented on lib/AST/GenericSignature.cpp:
```diff
@@ -1029,6 +1029,14 @@ unsigned GenericParamKey::findIndexIn(
+SubstitutionMap GenericSignature::getIdentitySubstitutionMap() const {
+  return SubstitutionMap::get(const_cast<GenericSignature*>(this),
+                              [](SubstitutableType *t) -> Type {
+                                return Type(cast<GenericTypeParamType>(t));
```
Do you mind doing a quick search to see if this utility function could be used anywhere else? I seem to recall writing similar things

slavapestov commented on lib/AST/Type.cpp:
```diff
@@ -2415,6 +2430,108 @@ NestedArchetypeType::NestedArchetypeType(const ASTContext &Ctx,
+}
+
+Optional<ProtocolConformanceRef>
+ReplaceOpaqueTypesWithUnderlyingTypes::operator()(CanType maybeOpaqueType,
```
This looks similar to the previous method -- can you factor anything out?

slavapestov commented on lib/SILGen/SILGenConvert.cpp:
```diff
@@ -775,8 +775,9 @@ ManagedValue SILGenFunction::emitExistentialErasure(
-                                      existentialTL.getLoweredType(),
-                                      conformances);
+                      existentialTL.getLoweredType(),
+                      cast<MetatypeType>(concreteFormalType).getInstanceType(),
```
concreteFormalType.getMetatypeInstanceType()

jckarter replied:
That doesn't give back a canonical type, so I'd have to coerce it back. Not sure that's any cleaner.

slavapestov commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -2525,9 +2525,109 @@ class VarDeclUsageChecker : public ASTWalker {
+      if (otherCandidate.second->isEqual(opaqueTypeInContext))
+        continue;
+        
+      if (!underlyingType->isEqual(otherCandidate.second)) {
```
If you really wanted to be fancy, use TypeJoin here

jckarter replied:
I was going to leave it up to evolution to decide whether we should try to join divergent return types. I can see the motivation, but it also seems like it could cover up mistakes just as easily.

slavapestov commented on lib/Sema/TypeCheckProtocolInference.cpp:
```diff
@@ -1966,8 +1966,9 @@ auto AssociatedTypeInference::solve(ConformanceChecker &checker)
         return None;
 
-      if (replacement->hasArchetype())
+      if (replacement->hasArchetype()) {
```
Unnecessary edit?

slavapestov commented on test/type/opaque.swift:
```diff
@@ -0,0 +1,225 @@
+func zang() -> __opaque C & P & Q {
+  return D()
+}
+func zung() -> __opaque AnyObject {
```
What about `AnyObject & P` or `Any & Q`

slavapestov commented on test/type/opaque.swift:
```diff
@@ -0,0 +1,225 @@
+
+// Invalid positions
+
+typealias Foo = __opaque P // expected-error{{'opaque' types are only implemented}}
```
Add a test for `typealias Foo : __opaque P` too maybe

slavapestov commented on test/type/opaque.swift:
```diff
@@ -0,0 +1,225 @@
+
+func blib() -> P & __opaque Q { return 1 } // expected-error{{'opaque' should appear at the beginning}}
+func blab() -> (P, __opaque Q) { return (1, 2) } // expected-error{{'opaque' types are only implemented}}
+func blob() -> (__opaque P) -> P { return { $0 } } // expected-error{{'opaque' types are only implemented}}
```
More fuzzing:

```
func foo<T : __opaque P>()
func foo<T>() where T == _opaque P
protocol Q : __opaque P
```

etc

What about `(__opaque P).self` in expression context

slavapestov commented on test/type/opaque.swift:
```diff
@@ -0,0 +1,225 @@
+// TODO: Should be valid
+
+let foo: __opaque P = 1 // FIXME expected-error{{'opaque' types are only implemented}}
+var computedFoo: __opaque P {  // FIXME expected-error{{'opaque' types are only implemented}}
```
What about subscripts?

And a read-only computed property (or subscript) should be allowed, no?

jckarter replied:
Yeah, I was planning to add support for read-only properties and subscripts too.

slavapestov commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1511,6 +1511,11 @@ ERROR(noreturn_not_supported,none,
+// Opaque return types
+ERROR(opaque_type_invalid_constraint,none,
+      "an 'opaque' type must specify only 'Any', protocols, and/or a "
+      "base class", ())
```
AnyObject too, technically...

slavapestov commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3216,6 +3221,18 @@ ERROR(trailing_closure_requires_parens,none,
 
+ERROR(opaque_type_no_underlying_type_candidates,none,
+      "function declares an opaque return type, but has no return statements "
+      "in its body from which to infer an underlying type", ())
```
This should be allowed IMHO:

```
func f() -> __opaque P { fatalError() }
```

Once you have resilience working here, you can compile the metadata accessor as 'unreachable' and never concretize the type in SILGen

jckarter replied:
It would be easier to support unreachable opaque types with the runtime support in place, because we could then leave the underlying type opaque during SILGen/IRGen since the underlying type is never reachable.

slavapestov commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3526,6 +3543,10 @@ ERROR(unsupported_convention,none,
 
+// Opaque types
+ERROR(unsupported_opaque_type,none,
+      "'opaque' types are only implemented for the declared type of properties or the return type of functions", ())
```
What about subscript element types?

natecook1000 replied:
Should this mention "read-only" for properties and subscripts?

slavapestov reviewed (commented):
Sorry, my review is not very substantiative. But at least I read through most of the code.

slavapestov commented:
You have a binary file 'bar' that got added by mistake

jckarter committed:
Respond to Slava's feedback

jckarter committed:
Mangle opaque types without underlying types as ErrorType for now.

jckarter committed:
Remove ResilienceExpansion from substOpaqueTypes for now.

It's currently meaningless, and it'll require thought to pass the correct value when it becomes
meaningful.

jckarter commented:
@swift-ci Please build toolchain macOS platform

jckarter commented:
@swift-ci Please build toolchain macOS platform

jckarter commented:
@swift-ci Please build toolchain

jckarter committed:
Sema: Don't allow properties with inferred types to infer opaque result types.

This prevents opaque result types from propagating nontrivially into other declarations' types,
which may be confusing and create implementation complexities.

jckarter committed:
Add test for deserializing opaque types across modules

jckarter commented:
@swift-ci Please build toolchain macOS platform

jckarter commented:
The real implementation 
https://github.com/apple/swift/pull/22072 can do everything this can now.
